### ISSUE TITLE
Allow Windows Magnifier to follow NVDA virtual cursor

### ISSUE THREAD

**Issue by CyrilleB79** [nvaccess/nvda#12539]
### Is your feature request related to a problem? Please describe.
As a visually impaired person, I am using NVDA in conjunction with Windows Magnifier.
That works quite smoothly in applications such as Windows Explorer or text editors, since the magnified view follows the caret or the focus.
However on the web, the magnified view does not follow the virtual cursor (browse mode cursor) and what is read by NVDA often cannot be seen on the screen, unless an element making the focus move is encountered (e.g. a link).
I did not test recently, but I think the same issue occurs with Zoomtext and probably with other magnification tools.

### Describe the solution you'd like
NVDA should inform the Magnifier of its virtual cursor's location so that the magnified view can follow it.
Note that Windows Magnifier follows the narrator's cursor on the web.

### Describe alternatives you've considered
None
### Additional context
N/A

## Notes
* Cc @leonardder maybe you have any pointer to deal with this issue. Especially, I guess that there may be some sort of API to inform the ATs of the location of the cursor.
* This issue may partially blocked by #11658. However this would already be beneficial to have the magnified view updated for each line/sentence. Cc @FalkoBabbage, @lukaszgo1 who commented in this other issue.

**Comment by leonardder** [comment]
The vision framework should offer you everything you need as long as there is an option to tell the windows magnifier the virtual cursor position (like with Narrator. However, I'm afraid this is a private API only Narrator is allowed to access, or it is the magnifier that accesses Narrator, I don't know.

**Comment by seanbudd** [comment]
Update - we are currently working on a prototype for this, with likely inclusion in 2025.2

Current basic engineering plan:

1. create a magnifier window in NVDA
sample: https://github.com/microsoftarchive/msdn-code-gallery-microsoft/blob/411c271e537727d737a53fa2cbe99eaecac00cc0/Official%20Windows%20Platform%20Sample/Magnification%20API%20sample/%5BC%2B%2B%5D-Magnification%20API%20sample/C%2B%2B/Windowed/MagnifierSample.cpp
2. make magnifier track virtual cursor.
Use NVDA highlighter and screen curtain as a base.
3. make magnifier track mouse cursor, add additional cursor context support
4. create a basic settings GUI to configure magnifier windowed type (docked vs tracking, cursors to follow, size)
5. Get feedback
6. DRY refactor - Highlighter and screen curtain contain a lot of common functionality with this proof concept which can be abstracted

**Comment by Adriani90** [comment]
Nice approach, thanks very much for this great information Sean.
I think that a separate config GUI is not really needed, this could be done in the vision settings pannel, amybe there could be just another grouping labelled "cursor tracking via magnifier" or something like this.

**Comment by k-kolev1985** [comment]
I'm also excited about this news. 😊

I have a question about the features. Will this implementation support text and image edge smoothing like the Windows Magnifier does? That helps a lot with readability of text at larger zoom levels.

**Comment by hwf1324** [comment]
That's really nice.

I do have a question though, how does this work on the top window? Because in one of the third party magnifiers I use (ZoomIt), there is a problem with the zoom being obscured by the top-loading window.

**Comment by seanbudd** [comment]
>I think that a separate config GUI is not really needed, this could be done in the vision settings pannel, amybe there could be just another grouping labelled "cursor tracking via magnifier" or something like this.

Yep that would be the plan

>I have a question about the features. Will this implementation support text and image edge smoothing like the Windows Magnifier does? That helps a lot with readability of text at larger zoom levels.

I'm not sure - I should have an answer to that soon, once the first step is done.
If the test magnifier window, which should use the Windows Magnifier API, supports it natively, then yes it should.
If we have to do custom processing withing NVDA to do this, then no, at least not in the initial release of this.

**Comment by seanbudd** [comment]
>I do have a question though, how does this work on the top window? Because in one of the third party magnifiers I use (ZoomIt), there is a problem with the zoom being obscured by the top-loading window.

Could you explain this problem some more? The magnifier window should always be top level, and should always magnify the top-level visible window. However there is no implementation yet, so we can't be sure about limitations or bugs. Right now we are building a proof of concept prototype to test the UX and find out what is possible.

**Comment by hwf1324** [comment]
> Could you explain this problem some more? The magnifier window should always be top level, and should always magnify the top-level visible window. However there is no implementation yet, so we can't be sure about limitations or bugs. Right now we are building a proof of concept prototype to test the UX and find out what is possible.

Zoom in when a top-level window exists on the screen, and the original top-level window will cover it before the magnifier window.

![图片](https://github.com/user-attachments/assets/10f0b717-7881-4084-9a71-5ec8cda6151e)
Use ZoomIt to magnify the start menu.

It can be observed that the original top window is simultaneously displayed as the original window after zooming in.

Use the Windows Magnifier to zoom in on the Start menu.
Zooming in for a screenshot is a bit trickier.

Hopefully, you won't run into this problem using the Accessibility API.

**Comment by seanbudd** [comment]
Thanks for explaining @hwf1324 

I have another question that is somewhat related. Do you mind telling me what you like about Zoomit? How does it compare to other magnifiers, and what made you choose it? I'm hoping this will be useful in designing our magnifier

**Comment by hwf1324** [comment]
| feature | Zoomit | Windows Magnifier |
| --- | --- | --- |
| Custom zoom shortcuts. (Shouldn't be an issue for NVDA.) | Yes | No |
| Multi-display support. (It depends on personal preference.) | Enlarges the contents of only one display, leaving the other displays as they are. (The disadvantage is that there is no mouse cursor when moving the mouse to other monitors.) | Zoom in on all displays as a whole. |
| performance. (I'm not sure.) | LiveZoom is probably achieved by intercepting the screen and zooming in, potentially consuming more resources. | As a native programme, the possibilities can be better. |
| Hidden to the notification area. (Shouldn't be an issue for NVDA.) | Yes | No |
| Other features (It may not matter to NVDA.) | Zoomit offers many other features, (e.g. Record, Draw, Break, etc.). But personally I basically only use LiveZoom to zoom in on the screen. | Read the contents of the screen aloud |

**Comment by seanbudd** [comment]
Thanks!

**Comment by Boumtchack** [comment]
Hello,

I've been working on an add-on for this I've just discovered it was a known issue. I was wondering if it was still a work in progress? 

So far I've managed to watch if the review/navigator have moved and if it does move the cursor to it (basically a copy of the "nvda+shift+m" + loop).

It's working until I use tab, the mouse and focus go out of the magnifier frame, and I have to manually move the mouse for it to work again.

Using windows magnifier.

**Comment by seanbudd** [comment]
@Boumtchack we are working on an integrated magnifier within NVDA. If you'd like to work on fixing the bug in this ticket, please open a PR

**Comment by Boumtchack** [comment]
Do you have any idea when it would be ready? I tried to look at your milestones and can't find any mention of it.

I've managed to code something that kinda work, but I feel I'm missing a lot of key feature to realy understand what is working and why.

Here's my code for now if anyone passing by would find it usefull for now.

```python 
import globalPluginHandler
from logHandler import log
from scriptHandler import script
import winUser
import api
import mouseHandler
import wx

class GlobalPlugin(globalPluginHandler.GlobalPlugin):
	pluginIsActive = False
	lastPositionX = 0
	lastPositionY = 0
	
	@script(
		gesture="kb:NVDA+shift+w"
	)
	def script_ToggleMouseMove(self, gesture):
			self.pluginIsActive = not self.pluginIsActive
			if self.pluginIsActive:
				log.info('-------Plugin ON-------')
				wx.CallLater(50, self.moveMouse)
			else:
				log.info('------Plugin OFF-----')

	def moveMouse(self):
		reviewPosition = api.getReviewPosition()
		try:
			reviewPositionStartPoint = reviewPosition.pointAtStart
		except (NotImplementedError, LookupError):
			reviewPositionStartPoint = None

		if (		
		reviewPositionStartPoint
		):
			x = reviewPositionStartPoint.x
			y = reviewPositionStartPoint.y
			
		else:
			navigatorObject = api.getNavigatorObject()
			try:
				(left, top, width, height) = navigatorObject.location
			except: 
				log.info("Navigator Object not found \n RETURN")
				return
			x = left + (width // 2)
			y = top + (height // 2)

		self.check_if_moved(x,y)
		
		if self.pluginIsActive:
			wx.CallLater(50, self.moveMouse)

	def event_gainFocus(self, obj, nextHandler):
		if self.pluginIsActive:
			wx.CallLater(50, self.moveMouse)
			nextHandler()

	def check_if_moved(self,x,y):
		if x != self.lastPositionX or y != self.lastPositionY:
			winUser.setCursorPos(x, y)
			mouseHandler.executeMouseMoveEvent(x, y)
			winUser.mouse_event(winUser.MOUSEEVENTF_MOVE,0,0,0,None)
			self.lastPositionX = x
			self.lastPositionY = y
		else: 
			return 
```

**Comment by seanbudd** [comment]
@Boumtchack - we generally do not and cannot provide estimates for projects.

**Comment by Boumtchack** [comment]
Hello again,

I've finnally managed to found a solution, it ***seems*** to work as intended.

The error I was getting was that `winUser.setCursorPos()` was not sending or that Windows was not receiving it.
I used `winUser.SendInput()` wich seems to do the trick.
I cleaned up a little the previous code so here it goes: 

```python
import globalPluginHandler
from scriptHandler import script
import winUser
import api
import ui
import mouseHandler
import wx


class GlobalPlugin(globalPluginHandler.GlobalPlugin):
    pluginIsActive = False
    lastPositionX = 0
    lastPositionY = 0
    timer = None
    
    @script(
        description=("Toggle on and off the mouse pointer following the current navigator object"),
		category="Mouse",
        gesture="kb:NVDA+shift+w"
    )
    def script_ToggleMouseMove(self, gesture):
        self.pluginIsActive = not self.pluginIsActive
        if self.pluginIsActive:
            self.timer = wx.CallLater(50, self.moveMouse)

    # This function is pretty much a copy of script_moveMouseToNavigatorObject() wich is called with NVDA+shift+m
    def moveMouse(self):
        reviewPosition = api.getReviewPosition()
        try:
            reviewPositionStartPoint = reviewPosition.pointAtStart
        except (NotImplementedError, LookupError):
            reviewPositionStartPoint = None

        if reviewPositionStartPoint:
            x = reviewPositionStartPoint.x
            y = reviewPositionStartPoint.y

        else:
            navigatorObject = api.getNavigatorObject()

            try:
                (left, top, width, height) = navigatorObject.location
            except:
                ui.message("Object has no location")
                return
            
            x = left + (width // 2)
            y = top + (height // 2)

        if x != self.lastPositionX or y != self.lastPositionY:          
            winUser.setCursorPos(x, y)
            mouseHandler.executeMouseMoveEvent(x, y)
            self.nudgeMouse()
            self.lastPositionX = x
            self.lastPositionY = y

        self.plugin_continue_if_active()
    
    def plugin_continue_if_active(self):
        if self.pluginIsActive:
            self.timer.Stop()
            self.timer = wx.CallLater(50, self.moveMouse)

    #This function send a deeper signal than winUser.setCursorPos(x, y) so no more out of frame cursor
    def nudgeMouse(self):
        inputs = []
        for dx in (1, -1):
            inputs.append(winUser.Input(type=0, ii=winUser.Input_I(mi=winUser.MouseInput(
                dx=dx, dy=0, mouseData=0, dwFlags=1, time=0, dwExtraInfo=None))))
        winUser.SendInput(inputs)

    def event_gainFocus(self, obj, nextHandler):
        self.plugin_continue_if_active()
        nextHandler()

```

**Comment by seanbudd** [comment]
@Boumtchack - is this something you'd like to integrate into NVDA?

**Comment by Boumtchack** [comment]
Hello,

I'm actually looking at the add-on submission guide, as I've found a way to make a magnifier from NVDA using the windows magnifier api, I have 2 version of it:
- Mouse is kept in the center of the screen
- Mouse move the screen when it's close to the borders

I can send the zip file of the first one if you are interested in it

Edit:
I opened an issue for it, if it's ok I'll open an other for the second version (bordered)

https://github.com/nvaccess/addon-datastore/issues/5522

**Comment by seanbudd** [comment]
@Boumtchack - I've sent you an email to schedule a meeting to discuss this in detail
